### PR TITLE
txn: fix the ttlmanager and cleanup logic for 1pc and async commit

### LINKED ISSUE #23331: The ddl test suite of stmtflow failed with 1pc enabled
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sh
stmtflow play ddl/add_column/addcolumn_createindex_complex.t.sql  --dsn 'root:@tcp(...)/test'
```
The stmtflow tool is located at [here](https://github.com/zyguan/tidb-test-util/tree/master/cmd/stmtflow) and the test suite can be found [here](https://github.com/pingcap/automated-tests/tree/master/ticases/transaction/tests/ddl).

### 2. What did you expect to see? (Required)

No error reported.

### 3. What did you see instead (Required)

One of DDL got stuck and the test finally timed out.

### 4. What is your TiDB version? (Required)

master (cad8e15db2da8caf2cc0e03b2caa9939e35bf6f0)

### PR DESCRIPTION
**Pull request by cfzjywxk**
### What problem does this PR solve?

Issue Number: close #23331 

Problem Summary:
There are two problems:

- The left pessimistic locks are not cleaned up using 1pc protocol to commit, blocking other concurrent transactions on these keys.
- The ttlManager is not closed, if the max ts caculation error is reported using async commit or 1pc, thus the left pessimistic locks could not be resolved by the ddl backfill worker. 

### What is changed and how it works?

What's Changed:
1. Do cleanup if the error is reported when 1pc protocol is used.
2. Close the `ttlManager` if the execution result of `twoPhaseCommitter` is error.

How it Works:

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

### Release note 

- No release note.

### CODE REVIEW

File: store/tikv/txn.go [hunk 1]
```diff
@@ -250,7 +250,11 @@ func (txn *KVTxn) Commit(ctx context.Context) error {
 	}
 	defer func() {
 		// For async commit transactions, the ttl manager will be closed in the asynchronous commit goroutine.
-		if !committer.isAsyncCommit() {
+		if committer.isAsyncCommit() || committer.isOnePC() {
+			if err != nil {
+				committer.ttlManager.close()
```
Comment (cfzjywxk) [inline store/tikv/txn.go:255]: Should we just use something like `pessimisticRollbackMutations` instead?

File: store/tikv/2pc.go [hunk 2]
```diff
@@ -920,6 +920,9 @@ func (c *twoPhaseCommitter) execute(ctx context.Context) (err error) {
 		if c.isOnePC() {
 			// The error means the 1PC transaction failed.
 			if err != nil {
+				if c.getUndeterminedErr() == nil {
+					c.cleanup(ctx)
```
Comment (cfzjywxk) [inline store/tikv/2pc.go:924]: Should we just use something like `pessimisticRollbackMutations` if it's pessimistic transactions instead of `cleanup`.

Comment (sticnarf) [reply store/tikv/2pc.go:924]: We can always use pessimistic rollback because a optimistic one pc won't leave any waste.

**Review by MyonKeminta — APPROVED** [review]
LGTM

**Comment by cfzjywxk** [comment]
@sticnarf @youjiali1995 PTAL

File: store/tikv/2pc.go (hunk 2 above)
Comment (youjiali1995) [reply store/tikv/2pc.go:924]: How about changing `cleanup()`?
```golang
		cleanupKeysCtx := context.WithValue(context.Background(), TxnStartKey, ctx.Value(TxnStartKey))
		var err error
		if c.isPessimistic && c.isOnePC() {
			err = c.pessimisticRollbackMutations(NewBackofferWithVars(cleanupKeysCtx, cleanupMaxBackoff, c.txn.vars), c.mutations)
		} else {
			err = c.cleanupMutations(NewBackofferWithVars(cleanupKeysCtx, cleanupMaxBackoff, c.txn.vars), c.mutations)

		}
		if err != nil {
```

**Review by youjiali1995 — COMMENTED** [review]
Rest LGTM

File: store/tikv/2pc.go (hunk 2 above)
Comment (sticnarf) [reply store/tikv/2pc.go:929]: @youjiali1995 We needn't do cleanup for optimistic 1PC, so I suggest this:

```go
if !c.isOnePC() {
	err = c.cleanupMutations(NewBackofferWithVars(cleanupKeysCtx, cleanupMaxBackoff, c.txn.vars), c.mutations)
} else if c.isPessimistic {
	err = c.pessimisticRollbackMutations(NewBackofferWithVars(cleanupKeysCtx, cleanupMaxBackoff, c.txn.vars), c.mutations)
}
```

Comment (youjiali1995) [reply store/tikv/2pc.go:929]: My mistake.

File: store/tikv/txn.go [hunk 3]
```diff
@@ -250,7 +250,11 @@ func (txn *KVTxn) Commit(ctx context.Context) error {
 	}
 	defer func() {
 		// For async commit transactions, the ttl manager will be closed in the asynchronous commit goroutine.
-		if !committer.isAsyncCommit() {
+		if committer.isAsyncCommit() || committer.isOnePC() {
```
Comment (cfzjywxk) [inline store/tikv/txn.go:253]: @sticnarf @youjiali1995 

Could we just always close the ttl manager here, seems at commit phase and the async commit/1pc transaction size will be small.

Comment (sticnarf) [reply store/tikv/txn.go:253]: I think it is okay. The chance is small that TTL expires before the primary lock is committed.

Comment (youjiali1995) [reply store/tikv/txn.go:253]: I think yes because `SafeWindow`(2s) is smaller than the interval of keepAlive(10s). It's nearly useless here.

Comment (sticnarf) [reply store/tikv/txn.go:253]: It seems to me not related to `SafeWindow`. `SafeWindow` only constrains the prewrite time.

File: store/tikv/2pc.go [hunk 4]
```diff
@@ -899,15 +899,20 @@ func (c *twoPhaseCommitter) cleanup(ctx context.Context) {
 		})
 
 		cleanupKeysCtx := context.WithValue(context.Background(), TxnStartKey, ctx.Value(TxnStartKey))
-		err := c.cleanupMutations(NewBackofferWithVars(cleanupKeysCtx, cleanupMaxBackoff, c.txn.vars), c.mutations)
+		var err error
+		if c.isPessimistic && c.isOnePC() {
```
Comment (sticnarf) [inline store/tikv/2pc.go:903]: How about skipping cleanup for optimistic one pc? https://github.com/pingcap/tidb/pull/23342/files#r596528085

**Review by youjiali1995 — COMMENTED** [review]
Should also remove https://github.com/pingcap/tidb/pull/23342/files#diff-7352f99d1adcf12eded39ff8934684296725268e9116dcecfa16f560827f9271R1173

**Comment by cfzjywxk** [comment]
> Should also remove https://github.com/pingcap/tidb/pull/23342/files#diff-7352f99d1adcf12eded39ff8934684296725268e9116dcecfa16f560827f9271R1173

@youjiali1995 
Which part?

**Comment by sticnarf** [comment]
> Should also remove https://github.com/pingcap/tidb/pull/23342/files#diff-7352f99d1adcf12eded39ff8934684296725268e9116dcecfa16f560827f9271R1173

I cannot see the line you refer to. I guess it's the `defer c.ttlManager.close()` in the asynchrounous commit stage?

**Comment by youjiali1995** [comment]
> I cannot see the line you refer to. I guess it's the `defer c.ttlManager.close()` in the asynchrounous commit stage?

Yes.

**Comment by cfzjywxk** [comment]
> > Should also remove https://github.com/pingcap/tidb/pull/23342/files#diff-7352f99d1adcf12eded39ff8934684296725268e9116dcecfa16f560827f9271R1173
> 
> I cannot see the line you refer to. I guess it's the `defer c.ttlManager.close()` in the asynchrounous commit stage?

Got, it's removed.

**Review by youjiali1995 — APPROVED** [review]

**Review by sticnarf — APPROVED** [review]
Lgtm

**Pull request merged**